### DISCUSSION TITLE
forge event with a json payload

### DISCUSSION THREAD

**Discussion by ouinouin** [arendst/Tasmota#16356] [Q&A]
Hi, i dont know if i m on the right way, i didn find anything related to this in the rules documentation (had a hard time to understand the event command.

on the console, i launch this command : 
`event powerstatus={"fullpowerpulled":2345 ,"ID":3}`
the goal is to parse the json in the rule to check the ID and make something with the powerpulled value.
unfortunately , the %value% inside the rule is not showing the payload.

  this doesnt work (the rule doesnt triggers) :
`ON event#powerstatus DO publish tests/result %value% ENDON"}} `

this works (meaning the trigger is working but i cannot forward the payload of the event inside the rule: 
`ON event#powerstatus DO publish tests/result "whateverhere" ENDON"}} `

any help on this subject ?

**Comment by sfromis** [comment]
Internally, the Event command generates some "invisible" JSON, which is why a regular JSON-based trigger works to pick up simple values from the event command. However, this breaks down when you mess with the JSON delimiters by using them yourself. This example works, as it is quite removed from being valid JSON:
`event powerstatus='fullpowerpulled':2345,'ID':3`

What you try to do would be straightforward with ESP32 and Berry, where you can just create your own command taking JSON arguments, and do what you like.

**Reply by barbudor** [reply]
Also a rule can only access to the value that triggered it
It can be the full JSON as a string (`on event#powerstatus do somethingwith %value% endon`) or a single field (`on event#powerstatus#ID do somethingwithID %value% endon`) But in the later case, you only get the ID, not the powervalue
THis is why, as suggested by sfromis, using Berry would be much more flexible

**Reply by sfromis** [reply]
The trouble is that it using classic rule triggers breaks down when a full JSON string is used as value for the Event command. With careful escaping and descaping in the Tasmota code, it could likely be made to work, but that's not how it is. It seems that any usage of `"` or `{` throws the Event command/trigger logic off track.

**Reply by barbudor** [reply]
using:
```
on event#test do var1 %value% endon
on event#test#b do var2 %value% endon
```
I get
```
18:03:21.941 CMD: event test={"a":4,"b":"hello"}
18:03:21.949 MQT: stat/dev-4119/EVENT = {"Event":"Done"}
18:03:22.045 RUL: EVENT#TEST performs "publish var1 "
18:03:22.050 MQT: var1 = 
18:03:22.063 RUL: EVENT#TEST#B performs "var2 HELLO"
18:03:22.069 MQT: stat/dev-4119/VAR = {"Var2":"HELLO"}
```
So the 2nd trigger works as expected but not the 1st

**Reply by sfromis** [reply]
Right, "the full JSON as a string" is just not available to a rule, which was the aim.

**Reply by ouinouin** [reply]
i got the point,  i d really like to switch to berry, but the logic i d like to build will be related to load shedding system using the multicast functionalities of Tasmota on some power monitoring sockets and other devices that have an esp8266.
i ll try to adapt my strategy upstream and send multiple events or something like that.
i never had a look at the scripting language on esp8266 , is it worth to spend some energy on it  ?

**Reply by sfromis** [reply]
The Scripting feature has a learning curve making it hard to get into, but you do get a lot of features, if you manage to do things just right. Hence, the number of people able to provide much help with issues is "low"..... I'm not using it, and do not know if it would help in this specific situation.

More generally, in the test example you use, you could of course "just" execute the Publish directly, without going via Event. Or you could refrain from using JSON as the data format, and make a payload format without those troublesome delimiters.
